### PR TITLE
Audition listing portal

### PR DESCRIPTION
**Pull request by hassanharoon86**

### CODE REVIEW

File: app/assets/stylesheets/application.scss [hunk 1]
```diff
@@ -10,6 +10,7 @@
  * files in this directory. Styles in this file should be added after the last require_* statement.
  * It is generally better to create a new file per style scope.
  *
+ /*
```
Comment (asad-imtiaz) [inline app/assets/stylesheets/application.scss:13]: why was this needed?

File: db/migrate/20211103053637_remove_assigned_to_from_auditions.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+class RemoveAssignedToFromAuditions < ActiveRecord::Migration[6.1]
+  def change
+    remove_column :auditions, :assigned_to, :string
```
Comment (asad-imtiaz) [inline db/migrate/20211103053637_remove_assigned_to_from_auditions.rb:3]: rollback the original migration and delete the file.

File: db/schema.rb [hunk 3]
```diff
@@ -10,11 +10,16 @@
 #
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema.define(version: 2021_10_29_053315) do
+ActiveRecord::Schema.define(version: 2021_11_03_054143) do
 
   # These are extensions that must be enabled in order to support this database
   enable_extension "plpgsql"
 
+  create_table "artists", force: :cascade do |t|
```
Comment (asad-imtiaz) [inline db/schema.rb:18]: please remove these tables in a migration and then remove that migration.

File: app/controllers/auditions_controller.rb [hunk 4]
```diff
@@ -13,6 +13,15 @@ def create
     end
   end
 
+  def update_assigned_to
+    @auditions = Audition.all.order(:id)
+    @audition = Audition.find(params[:audition_id])
+    @audition.user = User.find_by_email(params[:assigned_to] + '@audiosocket.com')
```
Comment (asad-imtiaz) [inline app/controllers/auditions_controller.rb:19]: use label and value in the select field.

File: app/controllers/home_controller.rb [hunk 5]
```diff
@@ -1,3 +1,9 @@
 class HomeController < ApplicationController
-  def index; end
+  def index
+    if user_signed_in?
```
Comment (asad-imtiaz) [inline app/controllers/home_controller.rb:3]: move in before action

File: app/controllers/managers_controller.rb [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    end
+
+    if params[:sorting_column].present?
+      @auditions = Audition.all.order(params[:sorting_column] + ' ' + params[:sorting_direction])
+    end
+  end
+
+  private
+
+  def manager_only
```
Comment (asad-imtiaz) [inline app/controllers/managers_controller.rb:18]: renameto verify_manger_user

File: app/controllers/managers_controller.rb [hunk 7]
```diff
@@ -0,0 +1,23 @@
+class ManagersController < ApplicationController
+  before_action :authenticate_user!, :manager_only
+
+  def index
+    @auditions = Audition.all.order(:id)
```
Comment (asad-imtiaz) [inline app/controllers/managers_controller.rb:5]: you can make a default scope to order

File: app/controllers/managers_controller.rb [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+    if params[:query].present?
+      @auditions = Audition.search(params[:query])
+    end
+
+    if params[:sorting_column].present?
+      @auditions = Audition.all.order(params[:sorting_column] + ' ' + params[:sorting_direction])
```
Comment (asad-imtiaz) [inline app/controllers/managers_controller.rb:12]: use  `params[:sorting_column] => params[:sorting_direction]`

File: app/controllers/managers_controller.rb [hunk 9]
```diff
@@ -0,0 +1,23 @@
+class ManagersController < ApplicationController
+  before_action :authenticate_user!, :manager_only
+
+  def index
+    @auditions = Audition.all.order(:id)
+
+    if params[:query].present?
```
Comment (asad-imtiaz) [inline app/controllers/managers_controller.rb:7]: you can use single line conditions here.

File: app/javascript/packs/manager.js [hunk 10]
```diff
@@ -0,0 +1,17 @@
+$(document).on('turbolinks:load', function() {
+  $('.manager-selection').on('change', function() {
+    $('#tr'+this.getAttribute('data-audition-id')).find("td:eq(5)").html(this.value)
+
+    var url_string = window.location.href;
+    var url = new URL(url_string);
+
+    $.ajax({
```
Comment (asad-imtiaz) [inline app/javascript/packs/manager.js:8]: try this `$.get('/path/to/action', param_one: value1, param_two: value2)`

File: app/models/audition.rb [hunk 11]
```diff
@@ -9,19 +11,32 @@ class Audition < ApplicationRecord
...
   SOURCES = ['Facebook', 'Instagram', 'Twitter', 'Other'].freeze
 
+  enum status: [:pending, :accepted, :rejected, :deleted]
+
   has_many :links, dependent: :destroy
+  belongs_to :user, optional: true
 
   accepts_nested_attributes_for :links, reject_if: :all_blank, allow_destroy: :true
 
+  validates :first_name, :last_name , length: { maximum: 30 },format: { with: NAME_REGEX, message: "No special characters allowed" }
```
Comment (asad-imtiaz) [inline app/models/audition.rb:21]: add space before `format:`

File: app/models/audition.rb [hunk 12]
```diff
@@ -9,19 +11,32 @@ class Audition < ApplicationRecord
...
   before_validation :remove_empty_genre
+  before_create :set_default_status
 
   private
 
   def remove_empty_genre
     genres.reject!(&:empty?)
   end
+
+  def set_default_status
```
Comment (asad-imtiaz) [inline app/models/audition.rb:39]: try to search how to set default enum value

File: app/controllers/auditions_controller.rb [hunk 13]
```diff
@@ -13,6 +13,15 @@ def create
     end
   end
 
+  def update_assigned_to
```
Comment (asad-imtiaz) [inline app/controllers/auditions_controller.rb:16]: I don't see a route defined for this action.

File: app/views/auditions/_audition.html.erb [hunk 14]
```diff
@@ -0,0 +1,29 @@
+<tr id="tr<%= audition.id %>">
+  <th scope="row"><%= audition.id %></th>
+
+  <td><%= audition.first_name + " " + audition.last_name %></td>
```
Comment (asad-imtiaz) [inline app/views/auditions/_audition.html.erb:4]: make a method called `full_name` in `Audition` model. and use string interpolation in that method.

File: app/views/auditions/_audition.html.erb [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+
+  <td><%= audition.artist_name %></td>
+
+  <td><%= audition.email %></td>
+
+  <td>
+    <% audition.genres.each do |genre| %>
+      <% if !(genre.equal?(audition.genres.last)) %>
```
Comment (asad-imtiaz) [inline app/views/auditions/_audition.html.erb:12]: try `genres.join(',')`

File: app/views/auditions/_audition.html.erb [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+      <% end %>
+    <% end %>
+  </td>
+
+  <td><%= audition.created_at.strftime("%d %b %Y %I:%M %p") %></td>
+
+  <% if audition.user_id == nil %>
+    <td><%= select_tag :assigned_to, options_for_select(User.where(role: 1).map{ |u| u.email.split('@').first }), class: 'form-control manager-selection', data: { audition_id: audition.id }, prompt: 'Select Manager' %></td>
+  <% else %>
+    <td><%= audition.user.email.split('@').first %></td>
```
Comment (asad-imtiaz) [inline app/views/auditions/_audition.html.erb:25]: create a method in user that returns the first part of its email and use that method.

File: app/controllers/auditions_controller.rb [hunk 17]
```diff
@@ -13,6 +13,14 @@ def create
...
+    @audition = Audition.find(params[:audition_id])
```
Comment (asad-imtiaz) [inline app/controllers/auditions_controller.rb:17]: move this in before_action

File: app/views/managers/_audition_table.html.erb [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+      <i class="fas fa-sort-down sort-color"></i>
+    <% elsif params[:sorting_column] == 'genres' && params[:sorting_direction] == 'DESC' %>
+      <i class="fas fa-sort-up sort-color"></i>
+    <% else %>
+      <i class="fas fa-sort sort-color"></i>
+    <% end %>
+  </th>
+
+  <th scope="col"><%= link_to 'SUBMITTED', sorting_column: 'created_at', sorting_direction: sortable('created_at'), remote: true %>
+    <% if params[:sorting_column] == 'created_at' && params[:sorting_direction] == 'ASC' %>
```
Comment (asad-imtiaz) [inline app/views/managers/_audition_table.html.erb:45]: move this logic into a helper method

File: db/migrate/20211103053637_remove_assigned_to_from_auditions.rb (hunk 2 above)
Comment (asad-imtiaz) [reply db/migrate/20211103053637_remove_assigned_to_from_auditions.rb:3]: ignore this.

Comment (hassanharoon86) [reply db/migrate/20211103053637_remove_assigned_to_from_auditions.rb:3]: ignored

File: app/controllers/auditions_controller.rb (hunk 13 above)
Comment (hassanharoon86) [reply app/controllers/auditions_controller.rb:18]: Just added one in routes.rb
